nuxeo-services/nuxeo-platform-usermanager-core/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java

```diff
@@ -1604,6 +1604,24 @@ public GroupConfig getGroupConfig() {
         return groupConfig;
     }
 
+    @Override
+    public List<String> getDescendantGroups(String groupId){
+        List<String> descendantGroups = new ArrayList<>();
+        populateDescendantGroups(groupId, descendantGroups);
+        return descendantGroups;
+    }
+
+    protected void populateDescendantGroups(String groupId, List<String> descendantGroups) {
+        NuxeoGroup group = getGroup(groupId);
+        if (group != null) {
+            List<String> subGroups = group.getMemberGroups();
+            subGroups.stream().filter(subGroup -> !descendantGroups.contains(subGroup)).forEach(subGroup -> {
+                descendantGroups.add(subGroup);
+                populateDescendantGroups(subGroup, descendantGroups);
+            });
+        }
+    }
```

Comment:
I can see 2 problems here, take the following group hierarchy (OK it's twisted, but can happen):
```
     A
     |--- B
     |    |--- A
     |    |--- C   
     |--- C
     |    |--- D
```
It seems that with the suggested algorithm, the resulting list will:
- Include A (the parent group, also sub-sub-group), but this is probably not an issue.
- Not include D, C being skipped since already present as a subgroup of B probably an issue.

Reply:
One solution could be to remove the duplicate filtering clause and use a HashSet (no duplicates) for the descendant groups.

Reply:
Yes maybe consider an algorithm based on (untested, modulo details):

        Set<String> descendantGroups = new HashSet<>();
        Deque<String> todo = new LinkedList<>();
        todo.add(groupId);
        while (!todo.isEmpty()) {
            String id = todo.pop();
            if (descendantGroups.add(id)) {
                // recurse
                NuxeoGroup group = getGroup(id);
                if (group != null) {
                    todo.addAll(group.getMemberGroups());
                }
            }
        }

Reply:
It seems to me that the actual algorithm works with this configuration...

Reply:
OK, then, can you please add a case in the test to show that we don't have the following problem:
> Not include D, C being skipped since already present as a subgroup of B probably an issue.

Reply:
Thanks, indeed it's working since when the first occurrence of C is added, its descendant groups are computed, thus D :+1: